### ISSUE TITLE
calling text::setOutlineThickness in a loop allocates a lot of memory (0.5 gigabyte)

### ISSUE THREAD

**Issue by DanielRabl** [SFML/SFML#1823]
I had a a lot more memory consumption than I should have had recently, so I went through every step and figured that calling `sf::Text::setOutlineThickness()` (therefore `sf::Text::ensureGeometryUpdate()`) in a loop with different values allocates a lot of memory (process memory reached about 550 megabytes after 10 seconds and stays there). 

Environment is Windows 10, Visual Studio 2019 (16.9.6). 
SFML version is 2.5.1 (Visual C++ 15 (2017) - 64-bit).
Build settings are Debug, x64, /std:c++latest.

This example should reproduce the behavior. Any font should work here (as I tried a few), here I used Arial font.

```cpp
#include <SFML/Graphics.hpp>

int main() {
	sf::RenderWindow window(sf::VideoMode(1280u, 720u), "SFML");

	sf::Font font;
	font.loadFromFile("arial.ttf");
	sf::Text text;
	text.setFont(font);
	text.setCharacterSize(60);
	text.setString("hello world!");
	text.setOutlineColor(sf::Color::White);

	sf::Clock time;
	float value = 0.0;
	while (window.isOpen()) {
		sf::Event event;
		while (window.pollEvent(event)) {
			if (event.type == sf::Event::Closed) window.close();
		}

		text.setOutlineThickness(value);
		value += 0.01;
		if (value > 10.0) {
			value = 0;
		}

		window.clear();
		window.draw(text);
		window.display();
	}
}
```
I do realize that calling `setOutlineThickness` this way is probably wrong as it's quite an expensive maneuver, however I don't expect there to be so much allocations. There should probably be a warning on the comment to `setOutlineThickness` if this memory problem is not to be solved.

Thanks!

**Comment by Bromeon** [comment]
I tried to reproduce it with VS 2019, latest SFML master (commit eeeda74ec1fd5de24343ee845af0d8cc89e992f6), Windows x64 Debug mode.

Could not see the behavior your experienced; memory stayed around ~36 MB and maybe 2-3 MB less with the `setOutlineThickness()` call commented out.

Did you set up a fresh project for this? Any other special configuration?

**Comment by DanielRabl** [comment]
Yes I made a completely fresh setup. I downloaded latest 2.5.1 (Visual C++ 15 (2017) - 64-bit version, linked and set include path, added the dlls and added arial.ttf, I added a Project.zip in the post.

**Comment by eXpl0it3r** [comment]
Changing the outline thickness, as well as changing any other visual properties, will trigger a recalculation of the geometry, which means that every glyph needs to be repositioned etc. This involves some `std::vector` clearing and appending.

The clear & append call shouldn't cause additional allocations, but I guess that's largely dependent on the STL implementation.
Since we're not doing anything magical with memory here, my guess is that this might just be some lazy deallocation or similar going on.

Have you tried this in release mode?

**Comment by Bromeon** [comment]
And out of interest, does it behave like this even if you don't set any text?

Asking because we have 
```cpp
 // No text: nothing to draw
if (m_string.isEmpty())
    return;
```
in the middle `Text::ensureGeometryUpdate()`, and that can bisect the potential cause 🙂

**Comment by DanielRabl** [comment]
Leaving the text empty will not cause the behavior.
Release mode has the exact same memory usage as in Debug mode.

**Comment by vittorioromeo** [comment]
Unless I am misunderstanding the code, the culprit seems to be `Font::getGlyph`, which does this:

```cpp
const Glyph& Font::getGlyph(Uint32 codePoint, unsigned int characterSize, bool bold, float outlineThickness) const
{
    // Get the page corresponding to the character size
    GlyphTable& glyphs = m_pages[characterSize].glyphs;

    // Build the key by combining the glyph index (based on code point), bold flag, and outline thickness
    Uint64 key = combine(outlineThickness, bold, FT_Get_Char_Index(static_cast<FT_Face>(m_face), codePoint));

    // Search the glyph into the cache
    if (auto it = glyphs.find(key); it != glyphs.end())
    {
        // Found: just return it
        return it->second;
    }
    else
    {
        // Not found: we have to load it
        Glyph glyph = loadGlyph(codePoint, characterSize, bold, outlineThickness);
        return glyphs.emplace(key, glyph).first->second;
    }
}
```

The `key` is dependent of `outlineThickness`, which varies on every iteration of your loop. Therefore we never find the glyph, and we have to load a new one with `loadGlyph`, then emplace it inside `glyphs` (memory allocation).

Why does the glyph depend on `outlineThickness`?

**Comment by Bromeon** [comment]
> Why does the glyph depend on `outlineThickness`?

It's this code in `loadGlyph()`:
https://github.com/SFML/SFML/blob/3579ecbdb040fbc6ee8f4222db28763a05b39bd9/src/SFML/Graphics/Font.cpp#L607-L613

[`FT_Stroker`](https://freetype.org/freetype2/docs/reference/ft2-glyph_stroker.html) is responsible of drawing the "stroke" (outline) onto the glyph.

---

So the cause of this issue is that for different outline thicknesses (and characters, character sizes, bold styles), separate glyphs are allocated. This is normally expected, because the user will use those glyphs to display text, and wants them to remain cached for fast rendering.

Now, the problem is that we never deallocate them. So cycling through many possible combinations of glyphs will cause lots of allocations, and memory that stays reserved. And the allocation pattern depends massively on the application:
1. Typical applications will use a handful of fonts with a handful of styles, so it's never a problem.
1. Others may use `sf::Text` heavily for graphical effects, needing many glyphs only for a a short time. This might be the case for some flashy games.
1. Yet other applications may be long-running, and over the course of their uptime request many different glyph patterns, and sometimes even render a huge number simultaneously. Text editors or complex UIs could fall into this category.

So unfortunately there's no easy way to garbage-collect glyphs. A possible strategy might be a LRU cache with a fixed size, but that will kill performance for applications that need more than the cache size simultaneously. Maybe exposing a tweakable strategy in the `sf::Font` API would be an approach, but it's very hard to find a good compromise between customizability and simplicity.

@DanielRabl could you tell us more about the project that made you notice this in the first place? (I.e. _not_ the reproducible example)

**Comment by DanielRabl** [comment]
I wanted to make a animation, where text is highlighted by a big outline thickness that vanishes quickly. Initially I reduced the outline thickness, after seeing the memory consumption I now instead just reduce the alpha value. It does make sense to me now why this memory is not cleared, but like I said therefore it's maybe worth warning about this behavior in the description of setOutlineThickness.

**Comment by Bromeon** [comment]
> I wanted to make a animation, where text is highlighted by a big outline thickness that vanishes quickly. Initially I reduced the outline thickness, [...]

So you started with a wide outline, and continuously made it narrower? 

Can you tell us roughly how many steps those were?

> [...]  it's maybe worth warning about this behavior in the description of setOutlineThickness.

Yes, that's the least. Even better would be if we could enable workflows like yours 🙂

**Comment by DanielRabl** [comment]
> Can you tell us roughly how many steps those were?
It was starting with a outline thickness of 2, and when the animation starts, it interpolates the value down to 0 every frame (144 frames / second) in the span of 1 second. For interpolation I used a curve slope (`return std::pow(1 - std::pow(1 - progress, slope), 1.0 / slope);`) where `slope` was 1.5 and progress is the time progress in range 0 to 1. But I believe it doesn't make a difference if it's a curve or if it's linear.

**Comment by winsvega** [comment]
Also, when setting 0 character size

**Comment by eXpl0it3r** [comment]
With the new `sf::Text` APIs that give you access to the glyph vertex data, I believe it's the better way to implement such effects, rather than requesting all possible outline thicknesses. As such, I'm closing this issue for now. If someone has a suggestions on reducing memory usage, feel free to continue the discussion here.

Look at the following new APIs added in SFML 3.1:
- `GlyphPreProcessor` / `setGlyphPreProcessor()`
- `getVertexData()`
- `getOutlineVertexData()`

Check the new [text example](https://github.com/SFML/SFML/blob/master/examples/text/Text.cpp) for usage examples.
